src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/edgpsi_gpu_ref.cl

```diff
@@ -0,0 +1,270 @@
+        if (p + 1 < h) {
+            if (top >= (kStackSize - 1)) {
+                FUNC_CALL(bubbleSortIterative)(arr, p + 1, h);
+            } else {
+                stack[++top] = p + 1;
+                stack[++top] = h;
+            }
+        }
+    }
+}
+
+// 1. Sort boxes by scores
+KERNEL(edgpsi_ref_stage_1)(__global INPUT0_TYPE* proposals) {
+    __global Box* boxes = (__global Box*)proposals;
+
+    FUNC_CALL(quickSortIterative)(boxes, 0, NUM_PROPOSALS-1);
+}
+#endif /* EDGPSI_STAGE_1 */
+
+#ifdef EDGPSI_STAGE_2
+
+// 2. NMS
+KERNEL(edgpsi_ref_stage_2)
+(const __global INPUT0_TYPE* boxes, __global size_t* out_indices, __global size_t* num_outputs) {
+    size_t count = 0;
+    size_t index_out[POST_NMS_COUNT] = {0};
+
+    size_t is_dead[PRE_NMS_TOPN] = {0};
+    for (size_t box = 0; box < PRE_NMS_TOPN; ++box) {
+        if (is_dead[box])
+            continue;
+
+        index_out[count++] = box;
+        if (count == POST_NMS_COUNT)
+            break;
+
+        const size_t box_offset = box * 5;
+        const float x0i = boxes[box_offset + 0];
+        const float y0i = boxes[box_offset + 1];
+        const float x1i = boxes[box_offset + 2];
+        const float y1i = boxes[box_offset + 3];
+
+        const float a_width = x1i - x0i;
+        const float a_height = y1i - y0i;
+        const float a_area = a_width * a_height;
+
+        for (size_t tail = box + 1; tail < PRE_NMS_TOPN; ++tail) {
+            const size_t tail_offset = tail * 5;
+            const float x0j = boxes[tail_offset + 0];
+            const float y0j = boxes[tail_offset + 1];
+            const float x1j = boxes[tail_offset + 2];
+            const float y1j = boxes[tail_offset + 3];
+
+            const float x0 = max(x0i, x0j);
+            const float y0 = max(y0i, y0j);
+            const float x1 = min(x1i, x1j);
+            const float y1 = min(y1i, y1j);
+
+            const float width = x1 - x0;
+            const float height = y1 - y0;
+            const float area = max(0.0f, width) * max(0.0f, height);
+
+            const float b_width = x1j - x0j;
+            const float b_height = y1j - y0j;
+            const float b_area = b_width * b_height;
+
+            const float intersection_area = area / (a_area + b_area - area);
+
+            is_dead[tail] =
+                (NMS_THRESHOLD < intersection_area) && (x0i <= x1j) && (y0i <= y1j) && (x0j <= x1i) && (y0j <= y1i);
+        }
+    }
+
+    *num_outputs = count;
+    for (size_t i = 0; i < count; ++i) {
+        out_indices[i] = index_out[i];
+    }
+}
+#endif /* EDGPSI_STAGE_2 */
+
+#ifdef EDGPSI_STAGE_3
+
+// 3. Convert proposals to rois and roi_scores
+KERNEL(edgpsi_ref_stage_3)
+(const __global INPUT0_TYPE* boxes,
+ const __global size_t* out_indices,
+ const __global size_t* num_outputs,
+ __global OUTPUT_TYPE* rois,
+ __global OUTPUT_TYPE* roi_scores) {
+    const size_t i = get_global_id(0);
+    const size_t index = out_indices[i];
+    const size_t box_offset = index * 5;
```

Comment:
No. All these calulations are not about offset within batch/feature/spatial dimensions. The thing is that input (box) and output (rois) are, in essense, not linear arrays, but rather arrays of structures:
    boxes: `{x1, y1, x2, y2, score}[0] .. {x1, y1, x2, y2, score}[N]`
    rois: `{x1, y1, x2, y2}[0] .. {x1, y1, x2, y2}[N]`
    
I do not see how `*_GET_INDEX` can be used here.

Reply:
I've used them in way
const uint output_idx = OUTPUT_GET_INDEX(b, 0, 0, 0);
const uint input_idx = INPUT0_GET_INDEX(roi_idx, 0, 0, 0);

potentially it's absolutely the same, maybe more flexible in case padding will be presented in future. But maybe better leave it as it is